noImplicitAny

When using `noImplicitAny: true` I get the following error:
```sh
inversify.d.ts(76,9): error TS7010: 'addPlan', which lacks return-type annotation, implicitly has an 'any' return type.
```

--- a/src/activation/decorator_utils.ts
+++ b/src/activation/decorator_utils.ts
@@ -1,8 +1,12 @@
 ///<reference path="../interfaces/interfaces.d.ts" />
 
+interface IReflectResult {
+    [key: string]: IMetadata[];
+}
+
 function tagParameter(target: any, targetKey: string, index: number, metadata: IMetadata) {
     let metadataKey = "inversify:tagged";
-    let paramsMetadata: Object = null;
+    let paramsMetadata: IReflectResult = null;
 
     // this decorator can be used in a constructor not a method
     if (targetKey !== undefined) {
@@ -37,12 +41,12 @@ function tagParameter(target: any, targetKey: string, index: number, metadata: I
     return target;
 }
 
-function _decorate(decorators, target) {
+function _decorate(decorators: ClassDecorator[], target: any): void {
     Reflect.decorate(decorators, target);
 }
 
-function _param(paramIndex, decorator) {
-    return function (target, key) { decorator(target, key, paramIndex); };
+function _param(paramIndex: number, decorator: ParameterDecorator): ClassDecorator {
+    return function (target: any, key?: string) { decorator(target, key, paramIndex); };
 }
 
 // Allows VanillaJS developers to use decorators:
@@ -56,9 +60,9 @@ function decorate(
     parameterIndex?: number): void {
 
     if (typeof parameterIndex === "number") {
-        _decorate([_param(parameterIndex, decorator)], target);
+        _decorate([_param(parameterIndex, <ParameterDecorator>decorator)], target);
     } else {
-        _decorate([decorator], target);
+        _decorate([<ClassDecorator>decorator], target);
     }
 }
 

--- a/src/interfaces/planning/context.d.ts
+++ b/src/interfaces/planning/context.d.ts
@@ -6,5 +6,5 @@ interface IContext {
         /// Gets or sets the activation plan.
         plan: IPlan;
 
-        addPlan(plan: IPlan);
+        addPlan(plan: IPlan): void;
 }

--- a/src/planning/planner.ts
+++ b/src/planning/planner.ts
@@ -50,7 +50,7 @@ class Planner implements IPlanner {
         try {
 
             let bindings = this.getBindings<any>(parentRequest.parentContext.kernel, target.service.value());
-            let activeBindings = [];
+            let activeBindings: IBinding<any>[] = [];
 
             if (bindings.length > 1 && target.isArray() === false) {
 
@@ -153,9 +153,9 @@ class Planner implements IPlanner {
         let paramNames = Reflect.getMetadata(METADATA_KEY.PARAM_NAMES, func) || [];
         let tags = Reflect.getMetadata(METADATA_KEY.TAGGED, func) || [];
 
-        let targets = injections.map((inject, index) => {
+        let targets = injections.map((injection: string, index: number) => {
             let targetName = paramNames[index];
-            let target = new Target(targetName, inject);
+            let target = new Target(targetName, injection);
             target.metadata = tags[index.toString()] || [];
             return target;
         });

--- a/src/resolution/resolver.ts
+++ b/src/resolution/resolver.ts
@@ -17,7 +17,7 @@ class Resolver implements IResolver {
         return this._inject(rootRequest);
     }
 
-    private _inject(request: IRequest) {
+    private _inject(request: IRequest): any {
 
         let bindings = request.bindings;
         let childRequests = request.childRequests;
@@ -29,7 +29,7 @@ class Resolver implements IResolver {
 
         } else {
 
-            let result = null;
+            let result: any = null;
             let binding = bindings[0];
             let isSingleton = binding.scope === BindingScope.Singleton;
 

--- a/test/activation/inject.test.ts
+++ b/test/activation/inject.test.ts
@@ -40,8 +40,11 @@ describe("@inject", () => {
 
     it("Should be usable in VanillaJS applications", () => {
 
+        interface IKatana {}
+        interface IShuriken {}
+
         let VanillaJSWarrior = (function () {
-            function VanillaJSWarrior(primary, secondary) {
+            function VanillaJSWarrior(primary: IKatana, secondary: IShuriken) {
                 // ...
             }
             return VanillaJSWarrior;

--- a/test/activation/named.test.ts
+++ b/test/activation/named.test.ts
@@ -1,5 +1,5 @@
-declare function __decorate(decorators, target, key?, desc?);
-declare function __param(paramIndex, decorator);
+declare function __decorate(decorators: ClassDecorator[], target: any, key?: any, desc?: any): void;
+declare function __param(paramIndex: number, decorator: ParameterDecorator): ClassDecorator;
 
 ///<reference path="../../typings/browser.d.ts" />
 
@@ -110,8 +110,11 @@ describe("@named", () => {
 
   it("Should be usable in VanillaJS applications", () => {
 
+    interface IKatana {}
+    interface IShurien {}
+
     let VanillaJSWarrior = (function () {
-        function NamedVanillaJSWarrior(primary, secondary) {
+        function NamedVanillaJSWarrior(primary: IKatana, secondary: IShurien) {
             // ...
         }
         return NamedVanillaJSWarrior;

--- a/test/activation/paramnames.test.ts
+++ b/test/activation/paramnames.test.ts
@@ -74,8 +74,11 @@ describe("@paramNames", () => {
 
     it("Should be usable in VanillaJS applications", () => {
 
+        interface IKatana {}
+        interface IShurien {}
+
         let VanillaJSWarrior = (function () {
-            function VanillaJSWarrior(primary, secondary) {
+            function VanillaJSWarrior(primary: IKatana, secondary: IShurien) {
                 // ...
             }
             return VanillaJSWarrior;

--- a/test/activation/tagged.test.ts
+++ b/test/activation/tagged.test.ts
@@ -1,5 +1,5 @@
-declare function __decorate(decorators, target, key?, desc?);
-declare function __param(paramIndex, decorator);
+declare function __decorate(decorators: ClassDecorator[], target: any, key?: any, desc?: any): void;
+declare function __param(paramIndex: number, decorator: ParameterDecorator): ClassDecorator;
 
 ///<reference path="../../typings/browser.d.ts" />
 
@@ -170,8 +170,11 @@ describe("@Tagged", () => {
 
   it("Should be usable in VanillaJS applications", () => {
 
+    interface IKatana {}
+    interface IShurien {}
+
     let VanillaJSWarrior = (function () {
-        function TaggedVanillaJSWarrior(primary, secondary) {
+        function TaggedVanillaJSWarrior(primary: IKatana, secondary: IShurien) {
             // ...
         }
         return TaggedVanillaJSWarrior;

--- a/test/inversify.test.ts
+++ b/test/inversify.test.ts
@@ -19,7 +19,7 @@ describe("InversifyJS", () => {
       }
 
       interface IShuriken {
-          throw();
+          throw(): string;
       }
 
       class Katana implements IKatana {
@@ -74,7 +74,7 @@ describe("InversifyJS", () => {
         }
 
         interface IShuriken {
-            throw();
+            throw(): string;
         }
 
         class Katana implements IKatana {
@@ -131,7 +131,7 @@ describe("InversifyJS", () => {
         }
 
         interface IShuriken {
-            throw();
+            throw(): string;
         }
 
         class Katana implements IKatana {
@@ -228,7 +228,7 @@ describe("InversifyJS", () => {
       }
 
       interface IShuriken {
-          throw();
+          throw(): string;
       }
 
       class Katana implements IKatana {
@@ -283,7 +283,7 @@ describe("InversifyJS", () => {
         }
 
         interface IShuriken {
-            throw();
+            throw(): string;
         }
 
         class Katana implements IKatana {
@@ -343,7 +343,7 @@ describe("InversifyJS", () => {
         }
 
         interface IShuriken {
-            throw();
+            throw(): string;
         }
 
         class Katana implements IKatana {
@@ -467,11 +467,11 @@ describe("InversifyJS", () => {
 
         let kernel = new Kernel();
         kernel.bind<INinja>("INinja").to(Ninja);
-        let log = [];
+        let log: string[] = [];
 
         kernel.bind<IKatana>("IKatana").to(Katana).proxy((katana) => {
             let handler = {
-                apply: function(target, thisArgument, argumentsList) {
+                apply: function(target: any, thisArgument: any, argumentsList: any[]) {
                     log.push(`Starting: ${new Date().getTime()}`);
                     let result = target.apply(thisArgument, argumentsList);
                     log.push(`Finished: ${new Date().getTime()}`);

--- a/test/kernel/kernel.test.ts
+++ b/test/kernel/kernel.test.ts
@@ -19,7 +19,7 @@ describe("Kernel", () => {
 
   it("Should be able to use middleware as configuration", () => {
 
-      let logger = (context) => {
+      let logger = (context: IContext) => {
           console.log(context);
       };
 
@@ -175,7 +175,7 @@ describe("Kernel", () => {
 
           public name: string;
 
-          public constructor(name) {
+          public constructor(name: string) {
               this.name = name;
           }
       }
@@ -260,7 +260,7 @@ describe("Kernel", () => {
 
           public name: string;
 
-          public constructor(name) {
+          public constructor(name: string) {
               this.name = name;
           }
       }
@@ -309,7 +309,7 @@ describe("Kernel", () => {
 
           public name: string;
 
-          public constructor(name) {
+          public constructor(name: string) {
               this.name = name;
           }
       }
@@ -318,7 +318,7 @@ describe("Kernel", () => {
 
           public name: string;
 
-          public constructor(name) {
+          public constructor(name: string) {
               this.name = name;
           }
       }

--- a/test/resolution/resolver.test.ts
+++ b/test/resolution/resolver.test.ts
@@ -956,11 +956,11 @@ describe("Resolver", () => {
 
         // This is a global for unit testing but remember 
         // that it is not a good idea to use globals
-        let timeTracker = [];
+        let timeTracker: string[] = [];
 
         kernel.bind<IKatana>(katanaId).to(Katana).proxy((katana) => {
             let handler = {
-                apply: function(target, thisArgument, argumentsList) {
+                apply: function(target: any, thisArgument: any, argumentsList: any[]) {
                     timeTracker.push(`Starting ${target.name} ${new Date().getTime()}`);
                     let result = target.apply(thisArgument, argumentsList);
                     timeTracker.push(`Finished ${target.name} ${new Date().getTime()}`);

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -8,10 +8,10 @@
         "experimentalDecorators": true,
         "emitDecoratorMetadata": true,
         "declaration": false,
-        "noImplicitAny": false,
+        "noImplicitAny": true,
         "removeComments": true,
         "noLib": false,
         "preserveConstEnums": true,
-        "suppressImplicitAnyIndexErrors": true
+        "suppressImplicitAnyIndexErrors": false
     }
 }

--- a/type_definitions/inversify/inversify-global-tests.ts
+++ b/type_definitions/inversify/inversify-global-tests.ts
@@ -98,7 +98,7 @@ module global_module_test {
 
     kernel.bind<IKatana>("IKatana").to(Katana).proxy((katanaToBeInjected: IKatana) => {
         let handler = {
-            apply: function(target, thisArgument, argumentsList) {
+            apply: function(target: any, thisArgument: any, argumentsList: any[]) {
                 console.log(`Starting: ${performance.now()}`);
                 let result = target.apply(thisArgument, argumentsList);
                 console.log(`Finished: ${performance.now()}`);

--- a/type_definitions/inversify/inversify-test.ts
+++ b/type_definitions/inversify/inversify-test.ts
@@ -105,7 +105,7 @@ module external_module_test {
 
     kernel.bind<IKatana>("IKatana").to(Katana).proxy((katanaToBeInjected: IKatana) => {
         let handler = {
-            apply: function(target, thisArgument, argumentsList) {
+            apply: function(target: any, thisArgument: any, argumentsList: any[]) {
                 console.log(`Starting: ${performance.now()}`);
                 let result = target.apply(thisArgument, argumentsList);
                 console.log(`Finished: ${performance.now()}`);

